File: api/src/Sudoku/Board.php
<?php

namespace App\Sudoku;

class Board
{
    private array $cells;

    private function __construct(array $cells)
    {
        $this->cells = $cells;
    }

    public static function empty(): self
    {
        $cells = [];
        for ($r = 0; $r < 9; $r++) {
            for ($c = 0; $c < 9; $c++) {
                $cells[$r][$c] = ['value' => 0, 'readOnly' => false];
            }
        }
        return new self($cells);
    }

    public static function fromArray(array $data): self
    {
        $cells = [];
        foreach ($data as $r => $row) {
            foreach ($row as $c => $cell) {
                $cells[$r][$c] = [
                    'value'    => (int)($cell['value'] ?? 0),
                    'readOnly' => (bool)($cell['readOnly'] ?? false),
                ];
            }
        }
        return new self($cells);
    }

    public function getValue(int $row, int $col): int
    {
        return $this->cells[$row][$col]['value'];
    }

    public function setValue(int $row, int $col, int $value): void
    {
        $this->cells[$row][$col]['value'] = $value;
    }

    public function isReadOnly(int $row, int $col): bool
    {
        return $this->cells[$row][$col]['readOnly'];
    }

    public function setReadOnly(int $row, int $col, bool $readOnly): void
    {
        $this->cells[$row][$col]['readOnly'] = $readOnly;
    }

    public function isSafe(int $row, int $col, int $num): bool
    {
        for ($i = 0; $i < 9; $i++) {
            if ($this->cells[$row][$i]['value'] === $num || $this->cells[$i][$col]['value'] === $num) {
                return false;
            }
        }
        $startRow = $row - $row % 3;
        $startCol = $col - $col % 3;
        for ($i = 0; $i < 3; $i++) {
            for ($j = 0; $j < 3; $j++) {
                if ($this->cells[$i + $startRow][$j + $startCol]['value'] === $num) {
                    return false;
                }
            }
        }
        return true;
    }

    public function toArray(): array
    {
        return $this->cells;
    }
}

File: api_py/sudoku/board.py
from __future__ import annotations


class Board:
    def __init__(self, cells: list[list[dict]]) -> None:
        self._cells = cells

    @classmethod
    def empty(cls) -> "Board":
        cells = [[{"value": 0, "readOnly": False} for _ in range(9)] for _ in range(9)]
        return cls(cells)

    @classmethod
    def from_array(cls, data: list[list[dict]]) -> "Board":
        cells = [
            [
                {
                    "value": int(cell.get("value", 0)),
                    "readOnly": bool(cell.get("readOnly", False)),
                }
                for cell in row
            ]
            for row in data
        ]
        return cls(cells)

    def get_value(self, row: int, col: int) -> int:
        return self._cells[row][col]["value"]

    def set_value(self, row: int, col: int, value: int) -> None:
        self._cells[row][col]["value"] = value

    def is_read_only(self, row: int, col: int) -> bool:
        return self._cells[row][col]["readOnly"]

    def set_read_only(self, row: int, col: int, read_only: bool) -> None:
        self._cells[row][col]["readOnly"] = read_only

    def is_safe(self, row: int, col: int, num: int) -> bool:
        for i in range(9):
            if self._cells[row][i]["value"] == num or self._cells[i][col]["value"] == num:
                return False
        start_row = row - row % 3
        start_col = col - col % 3
        for i in range(3):
            for j in range(3):
                if self._cells[start_row + i][start_col + j]["value"] == num:
                    return False
        return True

    def to_array(self) -> list[list[dict]]:
        return [
            [{"value": cell["value"], "readOnly": cell["readOnly"]} for cell in row]
            for row in self._cells
        ]
